filament/backend/src/vulkan/VulkanHandles.h

```diff
@@ -94,11 +94,46 @@ struct VulkanVertexBuffer : public HwVertexBuffer, VulkanResource {
             VulkanResourceAllocator* allocator, uint8_t bufferCount, uint8_t attributeCount,
             uint32_t elementCount, AttributeArray const& attributes);
 
+    ~VulkanVertexBuffer();
+
     void setBuffer(VulkanBufferObject* bufferObject, uint32_t index);
 
-    utils::FixedCapacityVector<VulkanBuffer const*> buffers;
+    inline VkVertexInputAttributeDescription const* getAttribDescriptions() {
+        return mInfo->attribDesc.data();
+    }
+
+    inline VkVertexInputBindingDescription const* getBufferDescriptions() {
+        return mInfo->bufferDesc.data();
+    }
+
+    inline VkBuffer const* getVkBuffers() const {
+        return mInfo->vkbuffers.data();
+    }
+
+    inline VkDeviceSize const* getOffsets() const {
+        return mInfo->offsets.data();
+    }
 
 private:
+    struct PipelineInfo {
+        template <typename type>
+        using Array = utils::FixedCapacityVector<type>;
+
+        PipelineInfo(size_t size)
+            : attribDesc(size, VkVertexInputAttributeDescription{}),
+              bufferDesc(size, VkVertexInputBindingDescription{}),
+              vkbuffers(size, VK_NULL_HANDLE),
+              offsets(size, 0),
+              attribToBufferIndex(size, -1) {}
+
+        Array<VkVertexInputAttributeDescription> attribDesc;
+        Array<VkVertexInputBindingDescription> bufferDesc;
+        Array<VkBuffer> vkbuffers;
+        Array<VkDeviceSize> offsets;
+        Array<int8_t> attribToBufferIndex;
```

Comment:
all of this actually looks like a `utils::StructureOfArray<>` -- if you used it, it would be a single allocation for all arrays which would be a bit more efficient. Might maybe be easier to use too, since it makes things appear as a struct.

Reply:
Switched to SoA.